src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -90,16 +88,23 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     };
 
                     var streamFeature = httpConnectionContext.ConnectionFeatures.Get<IQuicStreamFeature>();
-                    var streamId = streamFeature.StreamId;
-                    HighestStreamId = streamId;
 
                     if (!streamFeature.CanWrite)
                     {
+                        // Unidirectional stream
                         var stream = new Http3ControlStream<TContext>(application, this, httpConnectionContext);
                         ThreadPool.UnsafeQueueUserWorkItem(stream, preferLocal: false);
                     }
                     else
                     {
+                        // Keep track of highest stream id seen for GOAWAY
+                        var streamId = streamFeature.StreamId;
+
+                        lock (_sync)
+                        {
+                            HighestStreamId = streamId;
```

Comment:
Why does this have to be synchronized? It looks like this is only set in this loop. Also, is it actually possible for the current streamId to be smaller than a previously handled stream?

Reply:
> Also, is it actually possible for the current streamId to be smaller than a previously handled stream?

I'm not 100% certain, I'd have to ask.

Reply:
Issue for GOAWAY here: https://github.com/dotnet/aspnetcore/issues/18957